tests/test_diff_rigid.py

```diff
@@ -0,0 +1,216 @@
+    constraint_solver.backward(dL_dqacc)
+
+    # Fetch gradients of the input variables
+    dL_dM = ti_to_torch(constraint_solver.constraint_state.dL_dM)
+    dL_djac = ti_to_torch(constraint_solver.constraint_state.dL_djac)
+    dL_daref = ti_to_torch(constraint_solver.constraint_state.dL_daref)
+    dL_defc_D = ti_to_torch(constraint_solver.constraint_state.dL_defc_D)
+    dL_dforce = ti_to_torch(constraint_solver.constraint_state.dL_dforce)
+
+    ### Compute directional derivatives along random directions
+    FD_EPS = 1e-4
+    TRIALS = 100
+    errors = {}
+
+    def compute_dL_error(dL_dx, x_type):
+        dL_error = 0.0
+        for trial in range(TRIALS):
+            rand_dx = torch.randn_like(dL_dx)
+            rand_dx = torch.nn.functional.normalize(rand_dx, dim=0 if x_type in ["force", "aref", "efc_D"] else [0, 1])
+            if x_type == "mass":
+                # Make rand_dx symmetric
+                rand_dx = (rand_dx + rand_dx.transpose(0, 1)) * 0.5
+
+            dL = (rand_dx * dL_dx).sum()
+
+            input_force = init_input_force
+            input_aref = init_input_aref
+            input_efc_D = init_input_efc_D
+            input_jac = init_input_jac
+            input_mass = init_input_mass
+
+            # 1 * eps
+            if x_type == "force":
+                input_force = init_input_force + rand_dx * FD_EPS
+            elif x_type == "aref":
+                input_aref = init_input_aref + rand_dx * FD_EPS
+            elif x_type == "efc_D":
+                input_efc_D = init_input_efc_D + rand_dx * FD_EPS
+            elif x_type == "jac":
+                input_jac = init_input_jac + rand_dx * FD_EPS
+            elif x_type == "mass":
+                input_mass = init_input_mass + rand_dx * FD_EPS
+            lossP1 = compute_loss(input_mass, input_jac, input_aref, input_efc_D, input_force)
+
+            # -1 * eps
+            if x_type == "force":
+                input_force = init_input_force - rand_dx * FD_EPS
+            elif x_type == "aref":
+                input_aref = init_input_aref - rand_dx * FD_EPS
+            elif x_type == "efc_D":
+                input_efc_D = init_input_efc_D - rand_dx * FD_EPS
+            elif x_type == "jac":
+                input_jac = init_input_jac - rand_dx * FD_EPS
+            elif x_type == "mass":
+                input_mass = init_input_mass - rand_dx * FD_EPS
+            lossP2 = compute_loss(input_mass, input_jac, input_aref, input_efc_D, input_force)
+            dL_fd = (lossP1 - lossP2) / (2 * FD_EPS)
+
+            dL_error += (dL - dL_fd).abs() / max(dL.abs(), dL_fd.abs(), gs.EPS)
+
+        dL_error /= TRIALS
+        return dL_error
+
+    dL_dforce_error = compute_dL_error(dL_dforce, "force")
+    dL_daref_error = compute_dL_error(dL_daref, "aref")
+    dL_defc_D_error = compute_dL_error(dL_defc_D, "efc_D")
+    dL_djac_error = compute_dL_error(dL_djac, "jac")
+    dL_dM_error = compute_dL_error(dL_dM, "mass")
```

Comment:
I would recommend NOT defining an inline function `compute_dL_error` and just do a for loop. Besides, just evaluate the assert on-the-fly instead of computing all the errors first, i.e.:
```python
for dL_dx, x_type in ((dL_dforce, "force"), (dL_daref, "aref"), (dL_defc_D, "efc_D"), (dL_djac, "jac"), (dL_dM, "mass")):
    dL_error = 0.0
    for trial in range(TRIALS):
        rand_dx = torch.randn_like(dL_dx)
        [...]
    assert_allclose(dL_error, 0.0, rtol=rtol, err_msg=f"Error is too large for '{mass}'.")
```